File: lab-1/GraphAlgo/Java/EeeeGraphAlgo/src/main/java/eeee/dgar/Algos.java
package eeee.dgar;

import java.util.ArrayList;
import java.util.List;

public class Algos
{
    public static void BFS(Node start)
    {
        List<Node> queue = new ArrayList<>();
        List<Node> next = new ArrayList<>();

        queue.add(start);

        while (queue.size() > 0)
        {
            for (Node node: queue)
            {
                node.setIsVisited(true);

                for (Node child: node.children())
                {
                    if (!child.isVisited())
                    {
                        next.add(child);
                    }
                }
            }

            queue = next;
            next = new ArrayList<>();
        }
    }

    public static void DFS(Node start)
    {
        start.setIsVisited(true);

        for (Node child: start.children())
        {
            if (!child.isVisited())
            {
                DFS(child);
            }
        }
    }
}

File: lab-1/GraphAlgo/CSharp/EeeeGraphAlgo/EeeeGraphAlgo/Algos.cs
namespace EeeeGraphAlgo
{
    public class Algos
    {
        public static void BFS(Node start)
        {
            var queue = new List<Node>();
            var next = new List<Node>();
            
            queue.Add(start);

            while (queue.Count > 0)
            {
                foreach (var node in queue)
                {
                    node.IsVisited = true;
                    next.AddRange(node.Children.Where(child => !child.IsVisited));
                }

                queue = next;
                next = new List<Node>();
            }
        }

        public static void DFS(Node start)
        {
            start.IsVisited = true;

            foreach (var child in start.Children.Where(child => !child.IsVisited))
            {
                DFS(child);
            }
        }
    }
}